### PR TITLE
MAINT: Remove BaseEstimator

### LINKED ISSUE #12748: Type incompatibility between sklearn's API and `SlidingEstimator`
I wanted to pass a `SlidingEstimator` instance to sklearn's `cross_val_predict()`, but it gets red squiggly lines:
Screenshot 2024-07-25 at 09 17 41

The error message is:

> Argument of type "SlidingEstimator" cannot be assigned to parameter "estimator" of type "BaseEstimator" in function "cross_val_predict"
>   "SlidingEstimator" is incompatible with "BaseEstimator"

The reason is that `SlidingEstimator` doesn't inherit from sklearn's `BaseEstimator`, but from `mne.fixes.BaseEstimator`.

My current workaround is `cast()`ing the type:

```python
from typing import cast
from sklearn.base import BaseEstimator

predictions = cross_val_predict(
    cast(BaseEstimator, model),
    X,
    labels["cue"],
    cv=cv,
    n_jobs=-1,
)
```

### PR DESCRIPTION
**Pull request by larsoner**
Closes #12748

1. Un-nest `sklearn` stuff in `mne/decoding` *except* places that we use private attributes. Someday maybe we shouldn't use them at all, but I'll save that for later. In the meantime I'd rather not break *all* of `mne.decoding` if any of those imports fail in the future.
2. Remove now-unnecessary custom mixin classes.
3. Use `check_array` now that we require sklearn >= 1.2 (it's available as of 1.1)
4. The `cov` stuff didn't end up being too terrible -- just had to write a little helper for `get_params` / `set_params`

### CODE REVIEW

File: mne/decoding/ssd.py [hunk 1]
```diff
@@ -331,6 +330,11 @@ def apply(self, X):
         X = pick_patterns @ X_ssd
         return X
 
+    @property
+    def picks_(self):
+        """Backward compatible wrapper for picks."""
+        return self.picks
```
Comment (drammock) [inline mne/decoding/ssd.py:336]: do we want to deprecate `.picks_` or just keep it around forever?

**Review by drammock — APPROVED** [review]
Thanks for tackling this! changes look reasonable / straightforward, so I'm inclined to trust our test suite here. Just one question.

File: mne/decoding/ssd.py (hunk 1 above)
Comment (larsoner) [reply mne/decoding/ssd.py:336]: It's pretty trivial to keep it I think so probably not worth deprecating

**Comment by hoechenberger** [comment]
thanks for working on this, Eric!

File: mne/decoding/ssd.py [hunk 2]
```diff
@@ -125,9 +124,9 @@ def __init__(
                 "The signal band-pass must be within the noise "
                 "band-pass!"
             )
-        self.picks_ = _picks_to_idx(info, picks, none="data", exclude="bads")
+        self.picks = _picks_to_idx(info, picks, none="data", exclude="bads")
```
Comment (agramfort) [inline mne/decoding/ssd.py:127]: there was a reason for this. Technically you should be able to do:

est = SSD(info, ..., picks=picks1)
est.fit(...)
est.picks = picks2
est.fit(...)

and it should just work. Now I see that it's already broken. The gist is that you should be able to overwrite init param and do a fit with the same instance.

so the init should just do self.picks = picks
and self.picks_ should be set in the fit.

File: mne/decoding/ssd.py (hunk 1 above)
Comment (agramfort) [reply mne/decoding/ssd.py:336]: see comment above

**Review by agramfort — COMMENTED** [review]
just some discussion on picks_ vs picks
if sklearn becomes just required in mne.decoding and not any mne import I am ok with that. We do have a CI run without sklearn?

File: mne/decoding/ssd.py (hunk 2 above)
Comment (larsoner) [reply mne/decoding/ssd.py:127]: Hah, I think more than one of our estimators do not work this way. But I can probably at least fix it here. Will push a fix

**Comment by larsoner** [comment]
Got approval and latest commits were just getting things green so in it goes!

**Pull request merged**